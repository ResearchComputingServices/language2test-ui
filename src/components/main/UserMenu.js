import React, { useState } from 'react';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDown';
import PropTypes from 'prop-types';
import _ from 'lodash';

function UserMenu({ displayName, dropdowns }) {
    const [anchorEl, setAnchorEl] = useState(null);
    const handleOpen = event => setAnchorEl(event.currentTarget);
    const handleClose = () => setAnchorEl(null);
    dropdowns = _.compact(dropdowns);

    return (
        <div>
            <div
                className='user-menu'
                onClick={handleOpen}
                onKeyDown={() => _.noop()}
                role='menuitem'
                tabIndex={0}
            >
                <h2 className='display-name'>
                    {displayName}
                </h2>
                {!_.isEmpty(dropdowns) && <ArrowDropDownIcon className='arrow-dropdown' />}
            </div>
            {!_.isEmpty(dropdowns)
                && (
                    <Menu
                        anchorEl={anchorEl}
                        anchorOrigin={{
                            vertical: 'bottom',
                            horizontal: 'center',
                        }}
                        elevation={2}
                        getContentAnchorEl={null}
                        keepMounted
                        onClose={handleClose}
                        open={Boolean(anchorEl)}
                        transformOrigin={{
                            vertical: 'top',
                            horizontal: 'center',
                        }}
                    >
                        {_.map(dropdowns, (dropdown, index) => (
                            <MenuItem
                                key={index}
                                data-testid={`user-menu-item-${dropdown.title}`}
                                disabled={dropdown.disabled}
                                onClick={() => {
                                    _.isFunction(dropdown.handler) && dropdown.handler();
                                    handleClose();
                                }}
                            >
                                {dropdown.Icon}
                                <span className='pl-2'>
                                    {' '}
                                    {dropdown.title}
                                    {' '}
                                </span>
                            </MenuItem>
                        ))}
                    </Menu>
                )}
        </div>
    );
}

UserMenu.propTypes = {
    dropdowns: PropTypes.array,
    displayName: PropTypes.string.isRequired,
};

UserMenu.defaultProps = { dropdowns: [] };

export default UserMenu;
