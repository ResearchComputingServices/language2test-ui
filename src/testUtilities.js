/* eslint-disable import/prefer-default-export */
/* eslint-disable prefer-const */
/* eslint-disable react/prop-types */
/* eslint-disable react/no-children-prop */
import React from 'react';
import { shallow, mount as mountComponent } from 'enzyme';
import { Router } from 'react-router-dom';
import toJSON from 'enzyme-to-json';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import _ from 'lodash';
import historyService from './services/HistoryService';
import { reducers } from './redux/slices';

const mockStore = configureMockStore({ reducers });

function WrapperComponent({ propsFunc, children, Component }) {
    const props = propsFunc();
    return (
        <Component
            children={children}
            {...props}
        />
    );
}

export function createTest(Component, opts = {}, callback) {
    let { props, children, mount, state } = opts;
    if (_.isNil(props)) props = {};
    if (_.isNil(children)) children = null;
    if (mount) {
        return () => {
            state = state || {};
            const mockMath = Object.create(global.Math);
            mockMath.random = () => 0.5;
            global.Math = mockMath;
            const mountedTree = mountComponent((
                <Router history={historyService.getHistory()}>
                    <Provider store={mockStore({ ...state })}>
                        <Component {...props} >
                            {children}
                        </Component>
                    </Provider>
                </Router>));
            _.isFunction(callback) && callback(mountedTree);
        };
    }
    if (_.isFunction(props)) {
        return () => {
            const shallowTree = shallow(
                <WrapperComponent
                    Component={Component}
                    propsFunc={props}
                >
                    {children}
                </WrapperComponent>,
            ).dive();
            _.isFunction(callback) && callback(shallowTree);
        };
    }
    return () => {
        const shallowTree = shallow((
            <Component {...props} >
                {children}
            </Component>
        ));
        _.isFunction(callback) && callback(shallowTree);
    };
}

export function createSnapshotTest(Component, opts = {}) {
    return createTest(Component, opts, tree => expect(toJSON(tree)).toMatchSnapshot());
}
