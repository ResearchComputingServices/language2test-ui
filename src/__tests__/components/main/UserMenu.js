import React from 'react';
import { shallow } from 'enzyme';
import UserMenu from '../../../components/main/UserMenu';
import { createSnapshotTest } from '../../../testUtilities';

describe('UserMenu', () => {
    it('should match snapshot', createSnapshotTest(UserMenu, {
        props: {
            handleLogout: jest.fn(),
            displayName: '',
        },
    }));

    it('should handle button clicks without crashing', () => {
        const Component = shallow(<UserMenu
            displayName=''
            handleLogout={jest.fn()}
        />);
        Component.find('.user-menu')
            .simulate('click', { currentTarget: <div /> });
    });
});
