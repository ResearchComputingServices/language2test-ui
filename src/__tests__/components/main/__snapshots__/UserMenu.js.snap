// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserMenu should match snapshot 1`] = `
<div>
  <div
    className="user-menu"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="menuitem"
    tabIndex={0}
  >
    <h2
      className="display-name"
    />
  </div>
</div>
`;
